Created documentation of "Chords" (#51)
upsidedownlabs/upsidedownlabs.github.io: software/chords/index.rst

.. _chords:

Getting Started
#################

Overview
********
Chords is an open source web application designed for real-time signal visualization, particularly tailored for bio-potential signals like EEG, EMG, ECG and EOG. This tool serves as an advanced alternative to the standard Arduino IDE serial plotter, offering enhanced functionality for students, researchers and hobbyists alike who want to record biopotential data using  `BioAmp hardware <https://docs.upsidedownlabs.tech/hardware/index.html>`_.

> .. figure:: media/chords_landing_page.*
>     :align: center
>     :alt: chords landing page
> 

Reviewer: ```suggestion
.. figure:: media/chords_landing_page.*
    :align: center
    :alt: chords landing page
    
    chords landing page

```

> Features
> ********
> - **Connection**: Effortlessly compatible with BioAmp Hardware and automatically detects the ADC of the development board running the Chords application, simplifying the setup process and ensuring a smooth and efficient workflow from data acquisition to visualization.
> - **Real-time Visualization**: Experience smooth real-time visualization of incoming data rendered with `WebGL-Plot <https://github.com/danchitnis/webgl-plot>`_. The system ensures smooth signal display, helping you monitor the signal flow without interruptions or lag.
> - **Recording**: You can record data in CSV format indefinitely or set a timer to stop recording automatically when it expires.
> - **Download/Delete**: Easily download or delete recorded data with just a few clicks. Manage your data efficiently, allowing for quick access to relevant data or removal of unwanted data to free up storage.
> - **Zoom**: Zoom-in and zoom-out feature, gives you more control as now you can either zoom in to look at signal or zoom out for an overall view.
> - **Mode Toggle**: Offers both light and dark modes, enabling you to switch according to your preference for a more comfortable user experience.
> - **Pause/Play**: Includes options to freeze and resume data display, allowing users to pause the visualization for detailed examination and then continue data monitoring seamlessly.
> - **Channel**: Supports plotting of up to six channels of data simultaneously in real time where each stream is color coded differently for easy identification.
> - **Disconnect**: Terminate the connection to the development board with a simple click. This functionality ensures a hassle-free disconnection process when you’re done with data collection or visualization.
> 

Reviewer: ```suggestion
Features
********
- **Connection**: It is effortlessly compatible with BioAmp Hardware attached to an Arduino-compatible development board running the `Chords Arduino firmware <https://github.com/upsidedownlabs/Chords-Arduino-Firmware>`_, simplifying the setup process and ensuring a smooth and efficient workflow from data acquisition to visualization.
- **Real-time Visualization**: Experience smooth real-time visualization of incoming data rendered with `WebGL-Plot <https://github.com/danchitnis/webgl-plot>`_. The system ensures smooth signal display, helping you monitor the flow without interruptions or lag.
- **Recording**: You can record data in CSV format indefinitely or set a timer to stop recording automatically when it expires.
- **Download/Delete**: Easily download or delete recorded data with just a few clicks. Manage your data efficiently, allowing for quick access to relevant data or removal of unwanted data to free up storage.
- **Zoom**: Zoom-in and zoom-out feature, gives you more control as now you can either zoom in to look at signal or zoom out for an overall view.
- **Theme Toggle**: Offers both light and dark themes, enabling you to switch according to your preference for a more comfortable user experience.
- **Pause/Play**: Includes options to freeze and resume data display, allowing users to pause the visualization for detailed examination and then continue data monitoring seamlessly.
- **Channel**: Supports plotting of up to six channels of data simultaneously in real time where each stream is color-coded differently for easy identification.
- **Disconnect**: Terminate the connection to the development board with a simple click. This functionality ensures a hassle-free disconnection process when you’re done with data collection or visualization.

```

> Software Requirements
> *********************
> 
> * You have to install `Chords Arduino Firmware <https://github.com/upsidedownlabs/Chords-Arduino-Firmware>`_ using arduino IDE.
> * For now Chords is compatible with chromium based browsers. To ensure compatibility with supported browsers, see :ref:`chords-browsers-compatibility`.
> 

Reviewer: ```suggestion
Software Requirements
*********************

* You have to flash  `Chords Arduino Firmware <https://github.com/upsidedownlabs/Chords-Arduino-Firmware>`_ to your Arduino board using Arduino IDE.
* For now the Chords web application is only compatible with Chromium-based browsers. To check your browser compatibility, see :ref:`chords-browsers-compatibility` section of this page.

```

> Hardware Requirements
> *********************
> To use Chords, you need:
> 
> - Compatible Development Board : Arduino Uno R3 And R4 Minima
> - A USB cable (type depends on board)
> - BioAmp Hardware : 
> 
>   - `BioAmp EXG Pill <https://docs.upsidedownlabs.tech/hardware/bioamp/bioamp-exg-pill/index.html>`_
>   - `Muscle BioAmp Biscute <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-biscute/index.html>`_
>   - `Muscle BioAmp Candy <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-candy/index.html>`_
>   - `Muscle BioAmp Patchy <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-patchy/index.html>`_
>   - `Muscle BioAmp Shield <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-shield/index.html>`_
>   
> 

Reviewer: ```suggestion
Hardware Requirements
*********************
To use Chords, you need:

- Compatible Development Board : Arduino Uno R3 And R4 Minima
- A USB cable (type depends on board)
- BioAmp Hardware : 

  - `BioAmp EXG Pill <https://docs.upsidedownlabs.tech/hardware/bioamp/bioamp-exg-pill/index.html>`_
  - `Muscle BioAmp Biscute <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-biscute/index.html>`_
  - `Muscle BioAmp Candy <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-candy/index.html>`_
  - `Muscle BioAmp Patchy <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-patchy/index.html>`_
  - `Muscle BioAmp Shield <https://docs.upsidedownlabs.tech/hardware/bioamp/muscle-bioamp-shield/index.html>`_
  
```

Author: no change suggested, resolving this conversation.

> How to Use Chords
> *******************
> 
> Uploading the Code
> ------------------
> 
> 1. Connect the Arduino board to your laptop using the USB cable.
> 2. Copy and paste the `Arduino Firmware  <https://github.com/upsidedownlabs/Chords-Arduino-Firmware>`_ into the Arduino IDE.
> 3. Select your board under `Board > Arduino AVR` in the tools menu.
> 4. Choose the correct COM port in the tools menu (the one that disappears when you disconnect the board).
> 5. Upload the code and open Chords in your web browser.
> 

Reviewer: ```suggestion
How to Use Chords
*******************

Uploading the Code
------------------

1. Connect the Arduino board to your laptop using the USB cable.
2. Copy and paste the `Arduino Firmware  <https://github.com/upsidedownlabs/Chords-Arduino-Firmware>`_ into the Arduino IDE.
3. Select your board from `Board` section in the tools menu.
4. Choose the correct COM port in the tools menu (the one that disappears when you disconnect the board).
5. Click on the `Upload` button to flash your Arduino board.
```

- **Download/Delete**: Easily download or delete recorded data with just a few clicks. Manage your data efficiently, allowing for quick access to relevant data or removal of unwanted data to free up storage.
- **Zoom**: Zoom-in and zoom-out feature, gives you more control as now you can either zoom in to look at signal or zoom out for an overall view.
- **Theme Toggle**: Offers both light and dark themes, enabling you to switch according to your preference for a more comfortable user experience.
- **Pause/Play**: Includes options to freeze and resume data display, allowing users to pause the visualization for detailed examination and then continue data monitoring seamlessly.
- **Channel**: Supports plotting of up to six channels of data simultaneously in real time where each stream is color-coded differently for easy identification.
- **Disconnect**: Terminate the connection to the development board with a simple click. This functionality ensures a hassle-free disconnection process when you’re done with data collection or visualization.

Software Requirements

> *********************

Reviewer: We have extra blank line here.

        :alt: chords plus
        :align: left
        :width: 20px
        :height: 20px
     - Increase Channel
   * - .. figure:: media/chords_minus.*
        :alt: chords minus
        :align: left
        :width: 20px
        :height: 20px
     - Decrease Channel
   * - .. figure:: media/chords_light_mode.*
        :alt: chords light theme
        :align: left
        :width: 20px
        :height: 20px
     - Light Mode
   * - .. figure:: media/chords_dark_mode.*
        :alt: chords dark theme
        :align: left
        :width: 20px
        :height: 20px
     - Dark Mode
    
.. _chords-browsers-compatibility:

Browser Compatibility
*********************

The web application is compatible with the **Web Serial API**, which is essential for its functionality. Supported browsers include the latest versions of:

- **Google Chrome**
- **Microsoft Edge**
- **Opera**

If the user's browser does not support the Web Serial API, a message will inform them of the incompatibility, recommending the use of a supported browser.

For more information, refer to MDN Web Docs on the `Web Serial API <https://developer.mozilla.org/en-US/docs/Web/API/Web_Serial_API>`_.

Technologies Used

> *******************

Reviewer: Why do we have these extra blank lines?

- **Next.js**: A framework built on top of React.js.
- **Web Serial API**: API for serial communication with devices.
- **Tailwind CSS**: Utility-first CSS framework.
- **Shadcn UI**: Reusable and customizable components, built with Radix UI.
- **WebGL Plot**: Real-time charting library.
- **IndexedDB API**: Browser-based database for storing recorded data.

.. figure:: media/chords_tech_stack.*

>     :align: center
>     :alt: chords tech stack
> 
>   

Reviewer: We have to provide the title to every figure, I did the same for one other image above doing it again here. Please make sure to do it for other independent figures. You can only skip it for the figures you have put in that table.

```suggestion
    :align: center
    :alt: chords tech stack
    
    chords tech stack
```

Author: okay
